Show percentage through the file in the statusline/modeline

oldaccountdeadname opened the pull request:
Not quite sure if there's a proper name for this; here's a screenshot of the new behavior (see the "61%" in the modeline):

![screenshot-of-changes](https://user-images.githubusercontent.com/46805277/138544025-a88503c1-3502-445b-9b77-855bdcfb8fc2.png)

I personally like this, but I don't know if everyone does. Tell me if you'd prefer this to be an option in the config rather than hard-coded!

oldaccountdeadname committed:
extract `title` in current_buffer_status_line_data

This will make following changes (adding a position indicator)
moderately simpler.

oldaccountdeadname committed:
display the percentage through the file that the cursor is at

A modeline may look like, for instance:

	NORMAL  [17%] src/presenters/mod.rs

as opposed to the previous:

	NORMAL  src/presenters/mod.rs

This is a small & slightly opinionated change.

jmacdonald commented:
I like it! I have a couple of questions/suggestions:

* The percentage seems based on the cursor position, but it might be better to use the last visible line of the window. That would prevent the number from changing when the cursor is moved within the window, which feels distracting. I also tend to think of these indicators as answering "how far down in the document am I looking?", which aligns with the window. Any idea what other editors have landed on for this?
* Does the percentage still display if all of the buffer content fits on-screen? This relates to my earlier point: if it's meant to be an indicator of the how far down the window is looking into the buffer, we should hide it if the answer to that question is obvious.
* I kinda feel like it might be a better fit at the end of the line, since it's not _critical_ information, unlike the name or editor mode. What do you think?

oldaccountdeadname committed:
present statusline by left/right elements

The previous logic to display the statusline did so by right aligning
only the last element. This isn't sufficient if more than one element is
to appear on the right (e.g., a cursor position indicator). So, this
commit replaces the logic to first render the right elements as
compactly as possible, then render the first left elements as compactly
as possible, but expand out the last left element to fill the remaining
space.

Behavior is identical; this is purely an API change.

oldaccountdeadname committed:
display precentage through file on the right

This only applies to normal mode, and is a straightforward extension of
ff0dc68's modifications.

oldaccountdeadname committed:
only show cursor percentage indicator when file is larger than height

This modifies the display function in normal mode without changing the
logic to gather the percentage, and should make the statusline a bit
cleaner on small files.

lincolnauster commented:
> The percentage seems based on the cursor position, but it might be
> better to use the last visible line of the window. That would prevent
> the number from changing when the cursor is moved within the window,
> which feels distracting. I also tend to think of these indicators as
> answering "how far down in the document am I looking?", which aligns
> with the window. Any idea what other editors have landed on for this?

Hm, interesting. I hadn't thought of it that way. As for what other
editors do, Neovim looks to provide this percentage via the location of
the cursor (in src/nvim/buffer.c, line 4022, current ***@***.***),
though I'm not sure about anything else. What would you think about a
config toggle for it? I think this might just be a matter of personal
taste and/or what people are used to.

Though FWIW, I've been using this basically since I patched it in
(add8bbc), and I haven't found the constant changing distracting. The
indicator only pops up in normal mode, where the cursor can't really
change positions, so I personally like it as is given the extra
'precision' of the cursor. Happy to add in a config option if you think
it's a good idea, though.

> * Does the percentage still display if all of the buffer content fits
> on-screen? This relates to my earlier point: if it's meant to be an
> indicator of the how far down the window is looking into the buffer,
> we should hide it if the answer to that question is obvious.

Oh, I hadn't thought of that - done in fc43e26!

> * I kinda feel like it might be a better fit at the end of the line,
> since it's not _critical_ information, unlike the name or editor mode.
> What do you think?

Agreed! Done across d7fbd47 and ff0dc68

lincolnauster commented:
I rebased the commits to fix up the merge conflict. The hashes mentioned above are now out-of-date, but hopefully the log is clear enough - thanks!

oldaccountdeadname committed:
use same colors throughout statusline

The percentage indicator previously used the default colors, which was
in contrast to the rest of the statusline, which used focused.
